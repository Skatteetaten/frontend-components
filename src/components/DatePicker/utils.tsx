export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
