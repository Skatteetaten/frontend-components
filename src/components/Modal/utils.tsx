import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
