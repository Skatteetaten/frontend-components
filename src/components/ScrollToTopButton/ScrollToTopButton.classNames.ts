import { mergeStyleSets } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { PaletteProps } from '..';

export const getClassNames = () => {
  const palette = getTheme().palette as PaletteProps;
  const whiteBackground = palette.skeColor.white;
  const hoverBackground = palette.skeColor.lightBlue;
  const color = palette.skeColor.blue;
  const hoverColor = palette.skeColor.blackAlt;

  // @ts-ignore TODO
  return mergeStyleSets({
    topcontainer: {
      width: '100%',
      marginTop: '40px',
      zIndex: 2,
      visibility: 'hidden',
      selectors: {
        '@media (min-width: 1170px)': {
          position: 'fixed',
          top: 0,
          right: 0
        }
      }
    },
    container: {
      position: 'relative !important',
      margin: '0 auto',
      opacity: 0,
      transition: '0.2s',
      selectors: {
        '@media (min-width: 900px)': {
          maxWidth: '878px'
        },
        '@media (min-width: 1170px)': {
          maxWidth: '1151.9px'
        }
      }
    },
    box: {
      marginTop: 0,
      selectors: {
        '@media (min-width: 1170px)': {
          top: '80vh',
          position: 'absolute',
          right: 0
        }
      }
    },
    actionButton: {
      display: 'block',
      textDecoration: 'none !important',
      textAlign: 'center',
      margin: '24px auto',
      selectors: {
        '.ms-Button-flexContainer': {
          flexDirection: 'column'
        },
        ':hover .ms-Button-flexContainer > div': {
          backgroundColor: hoverBackground,
          color: hoverColor
        }
      }
    },
    icon: {
      fontSize: '40px',
      color: color,
      display: 'inline'
    },
    iconFixateContainer: {
      background: whiteBackground,
      borderRadius: '50%',
      height: '30px',
      width: '30px',
      position: 'absolute'
    },
    iconFixate: {
      position: 'absolute',
      top: '-0.5em',
      left: '-0.2em'
    },
    label: {
      fontSize: '13px',
      fontWeight: 700,
      textAlign: 'center',
      marginTop: '40px',
      backgroundColor: 'white !important',
      borderRadius: '10px',
      color: color,
      padding: '2px 15px'
    },
    vis: {
      visibility: 'visible',
      opacity: '1 !important',
      transform: 'translateY(-15px)'
    }
  });
};
