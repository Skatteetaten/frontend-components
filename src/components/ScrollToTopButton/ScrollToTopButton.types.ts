export interface ScrollToTopButtonProps {
  /** Tekst som vises sammen med icon som kan trykkes på for å scrolle til toppen */
  label?: string;
  /** Global attributt som må være unik for hele HTML dokumentet */
  id?: string;
  /** Overstyring av stiler */
  customClassNames?: {
    topContainer: string;
    container: string;
    box: string;
    button: string;
    label: string;
    icon: string;
    iconContainer: string;
  };
  /** Overstyring av bredde/plassering for store skjermbredder*/
  containerMaxWidth?: string;
}

export interface ScrollToTopButtonState {
  skjult: boolean;
}
