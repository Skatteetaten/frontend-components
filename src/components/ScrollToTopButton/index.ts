export { default } from './ScrollToTopButton';
export * from './ScrollToTopButton';
