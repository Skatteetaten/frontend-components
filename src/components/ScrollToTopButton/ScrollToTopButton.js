import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import ActionButton from '../ActionButton/ActionButton';
import Icon from '../Icon/Icon';

import { getClassNames } from './ScrollToTopButton.classNames';

export class ScrollToTopButton extends React.PureComponent {
  static propTypes = {
    /** Tekst som vises sammen med icon som kan trykkes på for å scrolle til toppen */
    label: PropTypes.string,
    /** Global attributt som må være unik for hele HTML dokumentet */
    id: PropTypes.string
  };

  static defaultProps = {
    label: 'Til toppen'
  };

  state = {
    skjult: true
  };

  sjekkSkjulKnapp = () => {
    const { skjult } = this.state;

    window.scrollY === 0
      ? !skjult && this.setState({ skjult: true })
      : skjult && this.setState({ skjult: false });
  };

  componentDidMount() {
    window.addEventListener('scroll', this.sjekkSkjulKnapp);
    this.sjekkSkjulKnapp();
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.sjekkSkjulKnapp);
  }

  render() {
    const { id, label } = this.props;

    const styles = getClassNames(this.props);

    const visKlasse = this.state.skjult ? '' : styles.vis;

    return (
      <div id={id} className={styles.topcontainer}>
        <div className={classnames(styles.container, visKlasse)}>
          <div className={styles.box}>
            <ActionButton
              className={styles.actionButton}
              ariaLabel={label}
              onClick={() => window.scrollTo(0, 0)}
            >
              <div className={styles.iconFixateContainer}>
                <div className={styles.iconFixate}>
                  <Icon className={styles.icon} iconName={'MoveUp'} />
                </div>
              </div>
              <div className={styles.label}>{label}</div>
            </ActionButton>
          </div>
        </div>
      </div>
    );
  }
}

export default ScrollToTopButton;
