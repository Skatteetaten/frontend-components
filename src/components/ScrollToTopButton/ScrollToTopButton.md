** ScrollToTopButton brukes til å gå til toppen av siden. **

```js noeditor
import MessageBar from '@skatteetaten/frontend-components/MessageBar';

<MessageBar>Rull nedover på siden for å se knappen.</MessageBar>;
```

```js
import ScrollToTopButton from '@skatteetaten/frontend-components/ScrollToTopButton';

<div>
  <ScrollToTopButton label={'Til toppen'} />
</div>;
```

```js noeditor beskrivelse

<h3>Går til toppen av siden</h3>
<p>
  ScrollToTopButton brukes når man har lange sider og det er nyttig å gå tilbake
  til toppen av siden. Ved skjermbredde større enn 1170px vises symbolet nede
  til høyre som overlay. Er skjermbredden mindre vises den sentrert.
</p>
```

```js noeditor uu
<h3>Tips</h3>

  <ul>
  <li>Må kunne brukes med tastatur og flytte tastaturfokus til toppen, ikke bare rulle.</li>
  <li>Nettleseren skal ikke bare rulle til toppen, men også tastaturfokus skal også flyttes til toppen.</li>
  <li>Skjermleserbrukere vil hovedsaklig bruke hurtigtaster for å flytte til toppen.</li>
  <li>Det er ikke noe krav å ha slik funksjonalitet</li>
  </ul>

<h3>Mest relevante WCAG-krav</h3>

  <ul>
  <li>2.1.1 A, Tastatur</li>
  </ul>

<h3>WAI-ARIA</h3>

  <ul>
  <li>Aria-hidden="true" brukes for skjule funksjonaliteten for skjermlesere.</li>
  <li>All annen aria-koding er overflødig pga. at hele knappen er skjult for skjermlesere.</li>
  </ul>
```
