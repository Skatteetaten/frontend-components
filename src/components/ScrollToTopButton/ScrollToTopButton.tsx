import classnames from 'classnames';
import * as React from 'react';
import ActionButton from '../ActionButton/ActionButton';
import Icon from '../Icon/Icon';
import { getClassNames } from './ScrollToTopButton.classNames';

interface ScrollToTopButtonProps {
  /** Tekst som vises sammen med icon som kan trykkes på for å scrolle til toppen */
  label?: string;
  /** Global attributt som må være unik for hele HTML dokumentet */
  id?: string;
}

interface ScrollToTopButtonState {
  skjult: boolean;
}

export class ScrollToTopButton extends React.PureComponent<
  ScrollToTopButtonProps,
  ScrollToTopButtonState
> {
  static defaultProps = {
    label: 'Til toppen'
  };

  state = {
    skjult: true
  };

  sjekkSkjulKnapp = () => {
    const { skjult } = this.state;

    window.scrollY === 0
      ? !skjult && this.setState({ skjult: true })
      : skjult && this.setState({ skjult: false });
  };

  componentDidMount() {
    window.addEventListener('scroll', this.sjekkSkjulKnapp);
    this.sjekkSkjulKnapp();
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.sjekkSkjulKnapp);
  }

  render() {
    const { id, label } = this.props;
    // @ts-ignore TODO
    const styles = getClassNames(this.props);

    const visKlasse = this.state.skjult ? '' : styles.vis;

    return (
      <div id={id} className={styles.topcontainer}>
        <div className={classnames(styles.container, visKlasse)}>
          <div className={styles.box}>
            <ActionButton
              className={styles.actionButton}
              ariaLabel={label}
              onClick={() => window.scrollTo(0, 0)}
            >
              <div className={styles.iconFixateContainer}>
                <div className={styles.iconFixate}>
                  <Icon className={styles.icon} iconName={'MoveUp'} />
                </div>
              </div>
              <div className={styles.label}>{label}</div>
            </ActionButton>
          </div>
        </div>
      </div>
    );
  }
}

export default ScrollToTopButton;
