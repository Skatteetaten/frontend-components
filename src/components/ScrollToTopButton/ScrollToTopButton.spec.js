import React from 'react';
import { mount } from 'enzyme';
import ScrollToTopButton from './ScrollToTopButton';
import { matches } from '../utils/test-utils';

function oppsettFullDOM(props) {
  return mount(<ScrollToTopButton {...props} />);
}

describe('ScrollToTopButton komponent', () => {
  it('matcher snapshot', () => {
    matches(<ScrollToTopButton />);
  });

  it('skal vise icon', () => {
    const wrapper = oppsettFullDOM();

    const icon = wrapper.find('i');

    expect(icon.prop('data-icon-name')).toEqual('MoveUp');
  });

  it('kan sette label', () => {
    const wrapper = oppsettFullDOM({
      label: 'Tulletittel'
    });

    const label = wrapper.find('button > span');

    expect(label.html()).toContain('>Tulletittel</div>');
  });
});
