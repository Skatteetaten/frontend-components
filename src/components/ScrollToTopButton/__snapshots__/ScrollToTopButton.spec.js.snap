// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTopButton komponent matcher snapshot 1`] = `
<div
  className="topcontainer-40"
>
  <div
    className="container-41"
  >
    <div
      className="box-42"
    >
      <button
        aria-label="Til toppen"
        className="ms-Button ms-Button--action ms-Button--command actionButton-50"
        data-is-focusable={true}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        role="button"
        type="button"
      >
        <span
          className="ms-Button-flexContainer flexContainer-51"
          data-automationid="splitbuttonprimary"
        >
          <div
            className="iconFixateContainer-45"
          >
            <div
              className="iconFixate-46"
            >
              <i
                aria-hidden={true}
                className="icon-59"
                data-icon-name="MoveUp"
              >
                ÔÅ†
              </i>
            </div>
          </div>
          <div
            className="label-47"
          >
            Til toppen
          </div>
        </span>
      </button>
    </div>
  </div>
</div>
`;
